File: LargestValueinEachTreeRow_DFS.java
// Time Complexity : O(n) --> where n is the number of nodes in the input tree 
// Space Complexity : O(n/2)
// Did this code successfully run on Leetcode (515): Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    // DFS Approach
    List<Integer> result;
    public List<Integer> largestValues(TreeNode root) {
        result = new ArrayList<>();
        
        // edge case
        if (root == null) return result;
        dfs(root, 0);
        return result;
    }
    
    private void dfs(TreeNode root, int level) {
        // base case
        if (root == null) return;
       
        // logic
        if (result.size() <= level) result.add(root.val);
        else result.set(level, Math.max(root.val, result.get(level)));
        
        dfs(root.left, level + 1);
        dfs(root.right, level + 1);
    }
}

File: largest_value_in_each_tree_row_dfs.py
"""
Problem: 515. Find Largest Value in Each Tree Row
Leetcode: https://leetcode.com/problems/find-largest-value-in-each-tree-row/
Solution:
    - DFS iterative solution using stack
    - do depth first search traversal
    - Add the children to the stack that are valid for the next round.
Time Complexity: O(n), where n is the number of nodes in tree
Space Complexity: O(h), where h is the height of the tree
Did this code successfully run on Leetcode : Yes
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def largestValues(self, root: TreeNode) -> List[int]:
        if not root:
            return []

        result = []
        # By default add root to the stack with depth zero
        stack = [(root, 0)]

        while stack:
            node, depth = stack.pop()

            if len(result) == depth:
                result.append(node.val)
            else:
                result[depth] = max(result[depth], node.val)

            if node.left:
                stack.append((node.left, depth + 1))
            if node.right:
                stack.append((node.right, depth + 1))
        return result